### PR TITLE
Add OpenSearchTable in flint core

### PR DESCRIPTION
**Pull request by penghuo**
### Description
1. Deprecated scroll api with search_after api.
2. Deprecated createReader function from FlintClient and add Table/OpenSearchIndexTable in FlintCore. OpenSearchIndexTable will be the high level interface of FlintCore.  Table will be high level interface of FlintCore which is independent of query engine.
3. Perf result

| SQL Query                                                                                                   | PR search_after on shard (ms) | PIT on Shard with Search_After (ms) |
|-------------------------------------------------------------------------------------------------------------|----------------|-------------------------------------|
| `SELECT COUNT(*) FROM dev.default.logs-181998`                                                                | 17021.0        | 18256.0                             |
| `SELECT COUNT(*) FROM dev.default.logs-181998 WHERE status <> 0`                                              | 18507.0        | 18247.0                             |
| `SELECT COUNT(*), AVG(size) FROM dev.default.logs-181998`                                                     | 18281.0        | 19951.0                             |
| `SELECT AVG(CAST(size AS BIGINT)) FROM dev.default.logs-181998`                                               | 19109.0        | 19719.0                             |
| `SELECT MIN(@timestamp), MAX(@timestamp) FROM dev.default.logs-181998`                                        | 18420.0        | 20459.0                             |
| `SELECT status, COUNT() FROM dev.default.logs-181998 WHERE status <> 0 GROUP BY status ORDER BY COUNT() DESC` | 19731.0        | 20686.0                             |

### Issues Resolved
https://github.com/opensearch-project/opensearch-spark/issues/494

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/FlintScanBuilder.scala [hunk 1]
```diff
@@ -6,22 +6,26 @@
...
     with SupportsPushDownV2Filters
     with Logging {
 
   private var pushedPredicate = Array.empty[Predicate]
 
+  lazy val scan = FlintScan(tables, schema, options, pushedPredicate)
+
   override def build(): Scan = {
-    FlintScan(table, schema, options, pushedPredicate)
+    scan
```
Comment (dai-chen) [inline flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/FlintScanBuilder.scala:28]: this must be lazy?

File: flint-spark-integration/src/test/scala/org/opensearch/flint/spark/covering/ApplyFlintSparkCoveringIndexSuite.scala [hunk 2]
```diff
@@ -33,6 +35,11 @@ class ApplyFlintSparkCoveringIndexSuite extends FlintSuite with Matchers {
   private val clientBuilder = mockStatic(classOf[FlintClientBuilder])
   private val client = mock[FlintClient](RETURNS_DEEP_STUBS)
 
+  /** Mock IRestHighLevelClient to avoid looking for real OpenSearch cluster */
+  private val clientUtils = mockStatic(classOf[OpenSearchClientUtils])
+  private val openSearchClient = mock[IRestHighLevelClient](RETURNS_DEEP_STUBS)
+  private val pitResponse = mock[CreatePitResponse]
```
Comment (dai-chen) [inline flint-spark-integration/src/test/scala/org/opensearch/flint/spark/covering/ApplyFlintSparkCoveringIndexSuite.scala:41]: Is PIT still used?

File: flint-core/src/main/scala/org/opensearch/flint/core/table/OpenSearchIndexTable.scala [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+   */
+  lazy val pageSize: Int = {
+    if (option.getScrollSize.isPresent) {
+      option.getScrollSize.get()
+    } else {
+      val docCount = indexStats.total().docs().count()
+      if (docCount == 0) {
+        maxResultWindow
+      } else {
+        val totalSizeBytes = indexStats.total().store().sizeInBytes
```
Comment (dai-chen) [inline flint-core/src/main/scala/org/opensearch/flint/core/table/OpenSearchIndexTable.scala:62]: Should we use primary store size here?

File: flint-core/src/main/scala/org/opensearch/flint/core/table/OpenSearchIndexTable.scala [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+  /**
+   * Creates a reader for the table. Not supported for OpenSearchIndexTable.
+   *
+   * @param query
+   *   The query string.
+   * @return
+   *   A FlintReader instance.
+   * @throws UnsupportedOperationException
+   *   if this method is called.
```
Comment (dai-chen) [inline flint-core/src/main/scala/org/opensearch/flint/core/table/OpenSearchIndexTable.scala:106]: What does this mean?

File: flint-core/src/main/scala/org/opensearch/flint/core/storage/FlintOpenSearchClient.java [hunk 5]
```diff
@@ -64,15 +61,23 @@ public class FlintOpenSearchClient implements FlintClient {
...
-  private final static Function<String, String> SHARD_ID_PREFERENCE =
-      shardId -> shardId == null ? shardId : "_shards:"+shardId;
-
   private final FlintOptions options;
 
   public FlintOpenSearchClient(FlintOptions options) {
     this.options = options;
   }
 
+  public static QueryBuilder queryBuilder(String query) throws IOException {
```
Comment (dai-chen) [inline flint-core/src/main/scala/org/opensearch/flint/core/storage/FlintOpenSearchClient.java:70]: Is it better to move to the new `Table`?

File: flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/config/FlintSparkConf.scala [hunk 6]
```diff
@@ -112,14 +112,19 @@ object FlintSparkConf {
...
 
   val SCROLL_DURATION = FlintConfig(s"spark.datasource.flint.${FlintOptions.SCROLL_DURATION}")
     .datasourceOption()
     .doc("scroll duration in minutes")
     .createWithDefault(String.valueOf(FlintOptions.DEFAULT_SCROLL_DURATION))
 
+  val SUPPORT_SHARD = FlintConfig(s"spark.datasource.flint.${FlintOptions.SUPPORT_SHARD}")
+    .datasourceOption()
+    .doc("indicate does index support shard or not")
+    .createWithDefault(String.valueOf(FlintOptions.DEFAULT_SUPPORT_SHARD))
```
Comment (dai-chen) [inline flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/config/FlintSparkConf.scala:126]: Is this new option for testing or some use case? Does this actually mean if the OpenSearch table is partitioned or not?

File: flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java [hunk 7]
```diff
@@ -121,6 +124,17 @@ public UpdateResponse update(UpdateRequest updateRequest, RequestOptions options
         return execute(OS_WRITE_OP_METRIC_PREFIX, () -> client.update(updateRequest, options));
     }
 
+  @Override
+  public IndicesStatsResponse stats(IndicesStatsRequest request) throws IOException {
+    return execute(OS_WRITE_OP_METRIC_PREFIX,
```
[the remark this answers is not in the document]
Comment (penghuo) [reply flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java:129]: Read op.
Done.

File: flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/FlintScanBuilder.scala (hunk 1 above)
Comment (penghuo) [reply flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/FlintScanBuilder.scala:28]: revert it, not necessary.

File: flint-core/src/main/scala/org/opensearch/flint/core/table/OpenSearchIndexTable.scala (hunk 4 above)
Comment (penghuo) [reply flint-core/src/main/scala/org/opensearch/flint/core/table/OpenSearchIndexTable.scala:106]: Removed

File: flint-spark-integration/src/test/scala/org/opensearch/flint/spark/covering/ApplyFlintSparkCoveringIndexSuite.scala (hunk 2 above)
Comment (penghuo) [reply flint-spark-integration/src/test/scala/org/opensearch/flint/spark/covering/ApplyFlintSparkCoveringIndexSuite.scala:41]: no, removed.

File: flint-core/src/main/scala/org/opensearch/flint/core/table/OpenSearchIndexTable.scala (hunk 3 above)
Comment (penghuo) [reply flint-core/src/main/scala/org/opensearch/flint/core/table/OpenSearchIndexTable.scala:62]: Done.

File: flint-core/src/main/scala/org/opensearch/flint/core/storage/FlintOpenSearchClient.java (hunk 5 above)
Comment (penghuo) [reply flint-core/src/main/scala/org/opensearch/flint/core/storage/FlintOpenSearchClient.java:70]: good point.
done.

File: flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/config/FlintSparkConf.scala (hunk 6 above)
Comment (penghuo) [reply flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/config/FlintSparkConf.scala:126]: This describes whether the OpenSearch index supports sharding. The reason is that OpenSearch serverless does not support sharding. 
Currently, a table's partition is determined by the partition, but there is no hard limit.

Comment (penghuo) [reply flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/config/FlintSparkConf.scala:132]: Ideally, we should remove this setting. However, currently, the AOSS index settings return shard information, so we cannot distinguish it.

```
{
    "settings": {
      "index": {
        "number_of_shards": "2",
        "number_of_replicas": "0",
      }
    }
  }
}
```

**Pull request merged**